src/Loop/Driver.php

```diff
@@ -285,4 +303,56 @@
      * @return null|object|resource The loop handle the event loop operates on. `null` if there is none.
      */
     abstract public function getHandle();
+
+    /**
+     * Create a watcher identifier.
+     *
+     * Any watcher identifiers returned to clients by the driver MUST be created using this method. The returned watcher
+     * identifier SHALL be unique for the duration of the PHP process.
+     *
+     * @return string The watcher identifier.
+     */
+    final protected function createWatcherIdentifier()
+    {
+        if (!isset($this->driverId)) {
+            $this->driverId = self::$nextDriverId++;
+        }
+
+        return "{$this->driverId}-" . $this->nextWatcherId++;
+    }
+
+    /**
+     * Validate a watcher identifier which was passed to the driver by a client.
+     *
+     * The driver SHOULD call this method when it is provided with a watcher identifier which it does not recognise. The
+     * driver SHOULD NOT call this method every time it receives a watcher identifier, as doing so would be relatively
+     * expensive and is not typically necessary.
+     *
+     * @param string $watcherId The watcher identifier.
+     * @param bool $mustThrow Whether the method MUST throw. For example, if `enable()` was called with an unrecognised
+     * watcher identifier then this should be set to `true`, but if `cancel()` was called it should be set to `false`.
+     *
+     * @throws InvalidWatcherException
+     */
+    final protected function validateWatcherIdentifier($watcherId, bool $mustThrow)
+    {
+        if (!\preg_match('{^(?P<driver_id>[a-z]+)-(?P<watcher_id>[a-z]+)$}', $watcherId, $matches)) {
```

Comment:
This function would only be called in error cases or when `cancel()` is called on a non-existent watcher. Is that still a problem? I guess the latter could be quite common?